klog.go

```diff
@@ -1041,13 +1102,23 @@ func (l *loggingT) setV(pc uintptr) Level {
 
 // Verbose is a boolean type that implements Infof (like Printf) etc.
 // See the documentation of V for more information.
-type Verbose bool
+type Verbose struct {
+	enabled bool
+	logr    logr.InfoLogger
+}
+
+func newVerbose(level Level, b bool) Verbose {
+	if logging.logr == nil {
+		return Verbose{b, nil}
+	}
+	return Verbose{b, logging.logr.V(int(level))}
+}
 
 // V reports whether verbosity at the call site is at least the requested level.
-// The returned value is a boolean of type Verbose, which implements Info, Infoln
+// The returned value is a struct of type Verbose, which implements Info, Infoln
 // and Infof. These methods will write to the Info log if called.
 // Thus, one may write either
-//	if glog.V(2) { glog.Info("log this") }
+//	if glog.V(2).Enabled() { glog.Info("log this") }
```

Comment:
I'm pretty sure there's code that relies on this being a bool, isn't there?

Reply:
Yep, there is.

There's not really any way we can do this cleanly without this, as a bool does not contain enough information to determine which V level should be logged at when `Info` is called on that V.

I mentioned it in the initial PR body, but the actual changeset needed to make kubernetes/kubernetes compile is:

```
 pkg/cloudprovider/providers/gce/cloud/gen.go                | 160 +++++++++++++++++++++++++++++++++++++++++++++++++--------------------------------------------------
 pkg/cloudprovider/providers/vsphere/vclib/connection.go     |   2 +-
 pkg/controller/nodelifecycle/node_lifecycle_controller.go   |   4 +--
 pkg/controller/volume/attachdetach/reconciler/reconciler.go |   6 ++--
 pkg/kubectl/cmd/util/helpers.go                             |   2 +-
 pkg/kubelet/eviction/helpers.go                             |   4 +--
 pkg/proxy/endpoints.go                                      |   2 +-
 pkg/scheduler/algorithm/predicates/predicates.go            |   8 ++---
 pkg/scheduler/algorithm/priorities/interpod_affinity.go     |   2 +-
 pkg/scheduler/algorithm/priorities/resource_allocation.go   |   2 +-
 pkg/scheduler/algorithm/priorities/resource_limits.go       |   2 +-
 pkg/scheduler/algorithm/priorities/selector_spreading.go    |   2 +-
 pkg/scheduler/core/generic_scheduler.go                     |   6 ++--
 plugin/pkg/auth/authorizer/rbac/rbac.go                     |   2 +-
 staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go  |   4 +--
 staging/src/k8s.io/apiserver/pkg/util/trace/trace.go        |   4 +--
 staging/src/k8s.io/client-go/rest/request.go                |   8 ++---
 staging/src/k8s.io/client-go/transport/round_trippers.go    |   8 ++---
 18 files changed, 114 insertions(+), 114 deletions(-)
```

(as of a couple of months ago).

I would be keen to hear any suggestions on how we can avoid this breaking change, but I fear we don't have many options if we want to preserve V levels 🙈

Reply:
Indeed, this is tricky.  Since we more or less adopted glog->klog, we sort of committed to not breaking it.

What about something like this:

make Verbose implement `logr.InfoLogger` and use the bool value to determine whether to do nothing or actually call the  glog method, which will call the logr method.

Ahh crud, they both define `Info()` with different signatures.

We could add a new `Vb()` method, fix our code, and leave `V()` users behind with some documented but wrong behavior.  It could even panic() if SetLogger was called.

Alternately - are there logr changes we could make to make bridging easier?

Alternatively, fork klog2 ?